### PR TITLE
Reimplement the shell widget

### LINKED ISSUE #3: Cursor for fullwidth characters
(in the tb-remote-ui branch) the text cursor is painted over existing characters. However for fullwidth characters, since we assume the cursor width is one cell it only covers half of the char.

snapshot-neovim-qt1

### LINKED ISSUE #84: White line on the top of window
I use `subtle` tiling WM. First screen - `vim-qt`, Second - `neovim-qt`

There is white line on the top of second screen. When I click right mouse button on it, then I can see small popup with inactive checkbox.
Also `nvim-qt` starts not fullscreen (left, top and bottom). Only reloading subtle's config makes `neovim-qt` fullscreen (but other programs starts fullscreen without reloading).

Arch Linux. latest neovim and neovim-qt

### PR DESCRIPTION
**Pull request by equalsraf**
This is a pretty big PR and no doubt will break some things. The code for this has been sitting in my hard drive for several months now and I've lacked the time to work it through, but we really need to push this to address a lot of outstanding issues.

- The new widget implementation is independent of the Neovim specific bits
- No more QImage, we only paint when needed
- The widget now has a copy of the contents (i.e. chars and formatting) that are displayed, this means we can redraw total or partial parts of the widget

### Challenges

1. Is performance worse? To me when testing it seems comparable, but I can't say for sure, specially on Windows (and I didn't get a chance to test on Mac). The store could likely benefit from manual optimizations.
2. Likely there are some regressions in rendering issues (off by a pixel, or something I've forgot to add)
3. Resizing is broken (but differently that before) but that is ok, we need to have a serious discussion on how resizing should behave - and with the previous implementations there were things we couldn't do. See #108.
4. I'm not very happy about the cell/shellcontents API

## Conditions

- [x] Fix #3
- [x] Hopefully this also addresses #39. **But I don't have a way to test with old Qt versions or a real HiDPI**.
- [x] Fix regression on #70, the timeout is being triggered, which is why it "seems" to work, we can connect to the neovimResized signal added in the last commit
- [x] Revert minimum size policy, we can't reduce the window size like this
- [x] Add --maximized not exactly the same as #83, but at least it seems to work without the previous issues
- [x]  Fix #84, or at least it should get rid of glitches seen in #75
- [x] BUG: when using the wildmenu, there is some sort of scrollling? bug

### CODE REVIEW

**Comment by equalsraf** [comment]
In Windows --maximized/--fullscreen is clearly racy, the window gets maximized but the actual Neovim size is smaller than what it should be.

**Comment by johanhelsing** [comment]
What kind of old qt versions so you wish to test with?

**Comment by equalsraf** [comment]
@johanhelsing not that important anymore since the behaviour last night seemed consistent in 5.5, but if you still have 5.4 around can you give it a try?

**Comment by johanhelsing** [comment]
Might give it a try when I get home, for the record, it looks nice with 5.5 except for a small white border at the top.

screenshot from 2016-03-15 10-55-20

**Comment by equalsraf** [comment]
:D @johanhelsing awesome. The while line is a margin from the docking area (#84), but maybe after this is done #75 will work too.

**Comment by equalsraf** [comment]
Is it just me or the wildmenu seems slow?

**Comment by johanhelsing** [comment]
Not sure if it's performance, but there's definitely some flickering going on when using wildmenu. Also, the cursor seems to jump around in places it shouldn't be.

It's also visible in some plugins, i.e. ctrl-p.

**Comment by equalsraf** [comment]
> Not sure if it's performance, but there's definitely some flickering going on when using wildmenu.

It almost seems that its failing to update the contents sometimes.

> Also, the cursor seems to jump around in places it shouldn't be.
.
Looking at the paint event it is possible we are painting the cursor in places where it is not really needed.

**Comment by fuelen** [comment]
[out-1.ogv.zip](https://github.com/equalsraf/neovim-qt/files/174708/out-1.ogv.zip)
Just see video in archive (subtle WM)

**Comment by equalsraf** [comment]
@fuelen not completely unexpected, saw a similar issue with i3. Just pushed a fix.

**Comment by fuelen** [comment]
It's look nice for now :)

File: src/gui/shell.cpp [hunk 1]
```diff
@@ -802,15 +615,31 @@ void Shell::wheelEvent(QWheelEvent *ev)
...
-			(cols != m_cols || rows != m_rows) ) {
-		m_nvim->neovimObject()->ui_try_resize(cols, rows);
+/// Resize remote Neovim.
+///
+/// Neovim ignores concurrent resizes. If you call this method while
+/// a resize is already in progress, the later call is delayed until
+/// the previous one is finished.
+void Shell::resizeNeovim(uint64_t n_cols, uint64_t n_rows)
+{
+	if (!m_nvim || (n_cols == columns() && n_rows == rows())) {
```
Comment (ivegotasthma) [inline src/gui/shell.cpp]: /home/ico/public/git/neovim-qt/src/gui/shell.cpp:636:25: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
      if (!m_nvim || (n_cols == columns() && n_rows == rows())) {
                             ^
    /home/ico/public/git/neovim-qt/src/gui/shell.cpp:636:48: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
      if (!m_nvim || (n_cols == columns() && n_rows == rows())) {
                                                    ^

Conversion warnings on Arch with gcc 5.3.0.

**Comment by equalsraf** [comment]
Perhaps related to the cursor issues https://github.com/equalsraf/neovim-qt/issues/112#issuecomment-197567229, thanks @fuelen, seems likely we are painting the cursor when we shouldn't.

**Comment by fuelen** [comment]
@equalsraf, oh sorry, that comment should be in this pull request

**Comment by equalsraf** [comment]
@fuelen pushed a fix for the cursor issue, can you check if it still happens?

**Comment by johanhelsing** [comment]
@equalsraf the cursor issue is fixed for me :)

**Comment by equalsraf** [comment]
@johanhelsing awesome, thanks.

Is anyone seeing performance issues on Linux, when scrolling (maybe with :set cul)

**Comment by fuelen** [comment]
@equalsraf and not fixed for me. 
And yes - navigation is very slow and jumping with :set cul

**Comment by equalsraf** [comment]
> And yes - navigation is very slow and jumping with :set cul

Pushed a fix for the performance issue.

**Comment by equalsraf** [comment]
Can't reproduce the cursor issue right now, @fuelen in the video what settings are you using for listchars/tabs/indentation?

**Comment by fuelen** [comment]
@equalsraf https://github.com/fuelen/nvimfiles

**Comment by equalsraf** [comment]
Thanks, :set cul was hiding the issue, should be fixed now.

**Comment by fuelen** [comment]
yes, fixed

**Comment by equalsraf** [comment]
Awesome, I still have to check this a bit further, but otherwise this should be ready to merge.

**Pull request merged**